<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Gota Morota" />

<meta name="date" content="2018-01-18" />

<title>Review of allele and genotypic frequencies</title>

<script src="day04_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="day04_files/bootstrap-3.3.6/css/bootstrap.min.css" rel="stylesheet" />
<script src="day04_files/bootstrap-3.3.6/js/bootstrap.min.js"></script>
<script src="day04_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<script src="day04_files/navigation-1.1/tabsets.js"></script>
<script src="day04_files/navigation-1.1/codefolding.js"></script>
<link href="day04_files/magnific-popup-1.1.0/magnific-popup.css" rel="stylesheet" />
<script src="day04_files/magnific-popup-1.1.0/jquery.magnific-popup.min.js"></script>
<link href="day04_files/readthedown-0.1/readthedown.css" rel="stylesheet" />
<script src="day04_files/readthedown-0.1/readthedown.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #ffffff; color: #1f1c1b; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; background-color: #ffffff; color: #a0a0a0; border-right: 1px solid #a0a0a0; }
td.sourceCode { padding-left: 5px; }
pre, code { color: #1f1c1b; background-color: #ffffff; }
code > span.kw { color: #1f1c1b; font-weight: bold; } /* Keyword */
code > span.dt { color: #0057ae; } /* DataType */
code > span.dv { color: #b08000; } /* DecVal */
code > span.bn { color: #b08000; } /* BaseN */
code > span.fl { color: #b08000; } /* Float */
code > span.cn { color: #aa5500; } /* Constant */
code > span.ch { color: #924c9d; } /* Char */
code > span.sc { color: #3daee9; } /* SpecialChar */
code > span.st { color: #bf0303; } /* String */
code > span.vs { color: #bf0303; } /* VerbatimString */
code > span.ss { color: #ff5500; } /* SpecialString */
code > span.im { color: #ff5500; } /* Import */
code > span.co { color: #898887; } /* Comment */
code > span.do { color: #607880; } /* Documentation */
code > span.an { color: #ca60ca; } /* Annotation */
code > span.cv { color: #0095ff; } /* CommentVar */
code > span.ot { color: #006e28; } /* Other */
code > span.fu { color: #644a9b; } /* Function */
code > span.va { color: #0057ae; } /* Variable */
code > span.cf { color: #1f1c1b; font-weight: bold; } /* ControlFlow */
code > span.op { color: #1f1c1b; } /* Operator */
code > span.bu { color: #644a9b; font-weight: bold; } /* BuiltIn */
code > span.ex { color: #0095ff; font-weight: bold; } /* Extension */
code > span.pp { color: #006e28; } /* Preprocessor */
code > span.at { color: #0057ae; } /* Attribute */
code > span.re { color: #0057ae; } /* RegionMarker */
code > span.in { color: #b08000; } /* Information */
code > span.wa { color: #bf0303; } /* Warning */
code > span.al { color: #bf0303; font-weight: bold; } /* Alert */
code > span.er { color: #bf0303; text-decoration: underline; } /* Error */
code > span. { color: #1f1c1b; } /* Normal */
</style>


</head>

<body>


<div id="content" data-toggle="wy-nav-shift">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->

<nav id="nav-top" role="navigation" aria-label="top navigation">
    <a role="button" href="#" data-toggle="wy-nav-top"><span class="glyphicon glyphicon-menu-hamburger"></span></a>
</nav>


<div id="header">
<h1 class="title">Review of allele and genotypic frequencies</h1>
</div>


<div id="table-of-contents">
    <h2><a href="#content">Review of allele and genotypic frequencies</a></h2>
    <div id="text-table-of-contents">
      <ul>
      <li><a href="#overview">Overview</a></li>
      <li><a href="#read-a-file">Read a file</a></li>
      <li><a href="#allele-frequency">Allele frequency</a></li>
      <li><a href="#genotypic-frequency">Genotypic frequency</a></li>
      <li><a href="#another-approach-for-obtaining-allele-frequency">Another approach for obtaining allele frequency</a></li>
      <li><a href="#compute-allele-frequencies-for-all-snps">Compute allele frequencies for all SNPs</a></li>
      <li><a href="#minor-allele-frequency">Minor allele frequency</a></li>
      <li><a href="#visualization">Visualization</a></li>
      <li><a href="#save-r-objects">Save R objects</a></li>
      </ul>
    </div>
</div>

<div id="main">
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>We will learn how to compute allele and genotypic frequencies in R using the cattle data set.</p>
</div>
<div id="read-a-file" class="section level2">
<h2>Read a file</h2>
<p>Use the function <code>read.table</code> to read the genotype file <code>Geno.txt</code> in a data frame format. We will store the genotype data in the variable <code>W</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">W &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">file =</span> <span class="kw">file.choose</span>(), <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>We can access a certain element in the data frame by entering its coordinate in the single square bracket <code>[]</code> operator. Let’s first access the element in the first column and the first row. When the row coordinate is omitted, the operator returns a data frame with just a single column.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">W[<span class="dv">1</span>, <span class="dv">1</span>]  <span class="co"># 1st row and 1st column </span>
<span class="kw">head</span>(W[, <span class="dv">1</span>])  <span class="co"># one-column data frame</span></code></pre></div>
<p>The following code shows the first five rows and columns.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">W[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</code></pre></div>
<p>We then drop the first column of data frame, which is the animal IDs. The <code>-</code> sign indicates dropping variables. So, <code>-1</code> means dropping the first column.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">W &lt;-<span class="st"> </span>W[, <span class="op">-</span><span class="dv">1</span>]</code></pre></div>
<p>Next, we will convert <code>W</code> into a matrix from a data frame. In R, matrices are more memory efficient and convenient than the other data types to do linear algebra.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">W &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(W)</code></pre></div>
<p>What is the dimension of <code>W</code>?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(W)</code></pre></div>
</div>
<div id="allele-frequency" class="section level2">
<h2>Allele frequency</h2>
<p>Let’s compute the allele frequency of the first SNP. The <code>table</code> function returns frequncies of genotypes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(W[, <span class="dv">1</span>])</code></pre></div>
<p>We can see that there are 100 <em>AA</em> animals, 475 <em>Aa</em> animals, and 429 <em>aa</em> animals. Let’s assign these numbers into variables.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nAA &lt;-<span class="st"> </span><span class="kw">table</span>(W[, <span class="dv">1</span>])[<span class="dv">3</span>]
nAa &lt;-<span class="st"> </span><span class="kw">table</span>(W[, <span class="dv">1</span>])[<span class="dv">2</span>]
naa &lt;-<span class="st"> </span><span class="kw">table</span>(W[, <span class="dv">1</span>])[<span class="dv">1</span>]</code></pre></div>
<p>Allele frequency of <em>A</em> is given by <span class="math display">\[ 
f(A) = p = \frac{2 \times (\text{no. of } AA \text{ individuals}) + 1 \times (\text{no. of } Aa \text{ individuals})}{2 \times \text{total no. of individuals}}.
\]</span></p>
<div id="exercise-1" class="section level3">
<h3>Exercise 1</h3>
<p>Use the variables <code>nAA</code>, <code>nAa</code>, and <code>naa</code> defined above and compute the allele frequencies of <em>A</em> and <em>a</em> in the first SNP.</p>
</div>
</div>
<div id="genotypic-frequency" class="section level2">
<h2>Genotypic frequency</h2>
<p>Genotypic frequency is given by <span class="math display">\[ 
f(AA) = P = \frac{\text{No. of } AA \text{ individuals}}{\text{Total no. individuals}} \\
f(Aa) = H = \frac{\text{No. of } Aa \text{ individuals}}{\text{Total no. individuals}} \\
f(aa) = Q = \frac{\text{No. of } aa \text{ individuals}}{\text{Total no. individuals}}. \\
\]</span></p>
<div id="exercise-2" class="section level3">
<h3>Exercise 2</h3>
<p>What are the genotypic frequencies of <code>AA</code>, <code>Aa</code>, and <code>aa</code> in the first SNP?</p>
</div>
</div>
<div id="another-approach-for-obtaining-allele-frequency" class="section level2">
<h2>Another approach for obtaining allele frequency</h2>
<p><span class="math display">\[ 
f(A) = p = \frac{2 \times (\text{frequency of } AA) + 1 \times (\text{frequency of } Aa)}{2 \times (\text{frequency of } AA + Aa + aa)}.
\]</span></p>
<div id="exercise-3" class="section level3">
<h3>Exercise 3</h3>
<p>Use the variables <code>P</code>, <code>H</code>, and <code>Q</code> defined above and compute the allele frequencies of <em>A</em> and <em>a</em> in the first SNP.</p>
</div>
<div id="exercise-4" class="section level3">
<h3>Exercise 4</h3>
<p>What are the genotypic frequencies of <code>AA</code>, <code>Aa</code>, and <code>aa</code> in the second SNP?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nAA &lt;-<span class="st"> </span><span class="kw">table</span>(W[, <span class="dv">2</span>])[<span class="dv">3</span>]
nAa &lt;-<span class="st"> </span><span class="kw">table</span>(W[, <span class="dv">2</span>])[<span class="dv">2</span>]
naa &lt;-<span class="st"> </span><span class="kw">table</span>(W[, <span class="dv">2</span>])[<span class="dv">1</span>]

p &lt;-<span class="st"> </span>(<span class="dv">2</span> <span class="op">*</span><span class="st"> </span>nAA <span class="op">+</span><span class="st"> </span><span class="dv">1</span> <span class="op">*</span><span class="st"> </span>nAa)<span class="op">/</span>(<span class="dv">2</span> <span class="op">*</span><span class="st"> </span>(nAA <span class="op">+</span><span class="st"> </span>nAa <span class="op">+</span><span class="st"> </span>naa))
p
q &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>p
q</code></pre></div>
</div>
</div>
<div id="compute-allele-frequencies-for-all-snps" class="section level2">
<h2>Compute allele frequencies for all SNPs</h2>
<p>So far we have learned how to compute the allele frequency of a single SNP using the <code>table</code> function. Here, we consider how to compute allele frequencies for the entire SNPs. Of course we can apply the <code>table</code> function manually one at a time. However, this approach takes too much time to compute allele frequencies for 6,960 SNPs. Recall that allele frequency of <em>A</em> is given by <span class="math display">\[ 
f(A) = p = \frac{2 \times (\text{no. of } AA \text{ individuals}) + 1 \times (\text{no. of } Aa \text{ individuals})}{2 \times \text{total no. of individuals}}.
\]</span> We can rewrite this equation into <span class="math display">\[ 
f(A) = p = \frac{(\text{no. of } A \text{ allele in the population})}{2 \times \text{total no. of individuals}}.
\]</span> This suggests that all we need is the number of <span class="math inline">\(A\)</span> allele or reference allele <span class="math inline">\(a\)</span> for each SNP. The <code>sum</code> function returns the number of reference allele <span class="math inline">\(A\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sum</span>(W[, <span class="dv">1</span>])  <span class="co"># sum of A allele in the first SNP</span>
<span class="kw">sum</span>(W[, <span class="dv">2</span>])  <span class="co"># sum of A allele in the second SNP</span></code></pre></div>
<p>How to repeat this operation for 6,960 SNPs? The <code>colSums</code> function returns the sum of each column in a matrix as a vector.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colSums</span>(W)</code></pre></div>
<p>Note that <code>colSums(W)</code> gives the numerator of the above equation. We then divide this number by <span class="math inline">\(2 \times \text{total no. of individuals}\)</span>. The function <code>nrows</code> returns the number of rows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">colSums</span>(W)<span class="op">/</span>(<span class="dv">2</span> <span class="op">*</span><span class="st"> </span><span class="kw">nrow</span>(W))</code></pre></div>
<p>The variable <code>p</code> is a vector and it contains the allele frequencies of reference allele for 6,960 SNPs.</p>
<div id="exercise-5" class="section level3">
<h3>Exercise 5</h3>
<p>What is the allele frequency of reference allele in the 300th SNP?</p>
</div>
<div id="exercise-6" class="section level3">
<h3>Exercise 6</h3>
<p>What is the mean of reference allele frquencies in this population?</p>
</div>
</div>
<div id="minor-allele-frequency" class="section level2">
<h2>Minor allele frequency</h2>
<p>In most cases, people report a minor allele frequency, which is the frequency of less frequent allele in a given SNP. We can convert allele frequencies into minor allele frquencies by using the <code>ifelse</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">maf &lt;-<span class="st"> </span><span class="kw">ifelse</span>(p <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dv">1</span> <span class="op">-</span><span class="st"> </span>p, p)</code></pre></div>
<div id="exercise-7" class="section level3">
<h3>Exercise 7</h3>
<p>What is the minor allele frquency of reference allele in the 300th SNP?</p>
</div>
<div id="exercise-8" class="section level3">
<h3>Exercise 8</h3>
<p>What is the mean of minor allele frquencies?</p>
</div>
</div>
<div id="visualization" class="section level2">
<h2>Visualization</h2>
<p>Now let’s visualize the minor allele frequencies for the first 500 SNPs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(maf[<span class="dv">1</span><span class="op">:</span><span class="dv">500</span>])</code></pre></div>
</div>
<div id="save-r-objects" class="section level2">
<h2>Save R objects</h2>
<p>Save the variable <code>W</code> so that we can reuse it in the next class.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">save</span>(W, <span class="dt">file =</span> <span class="st">&quot;W.Rda&quot;</span>)</code></pre></div>
</div>
</div>


</div>

<div id="postamble" data-toggle="wy-nav-shift" class="status">
<p class="author"><span class="glyphicon glyphicon-user"></span> Gota Morota</p>
<p class="date"><span class="glyphicon glyphicon-calendar"></span> January 18, 2018</p>
</div>


<script>
$(document).ready(function () {
 	  });
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
